app/services/price_policies/time_based_price_calculator.rb

```diff
@@ -40,6 +45,58 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
       end
     end
 
+    def duration_pricing_cost_and_subsidy(duration_mins)
+      duration_in_hours = duration_mins / 60.0
+
+      result = build_intervals.reduce({ time_left: duration_in_hours, cost: 0, subsidy: 0 }) do |acc, interval_data|
+        time_left = acc[:time_left]
+
+        interval_length = interval_data[:interval_end] - interval_data[:interval_start]
+
+        time_to_charge = [time_left, interval_length].min
+
+        acc[:time_left] -= time_to_charge
+        acc[:cost] += interval_data[:step_rate] ? interval_data[:step_rate] * time_to_charge : 0
+        acc[:subsidy] += interval_data[:step_subsidy] ? interval_data[:step_subsidy] * time_to_charge : 0
```

Comment:
```suggestion
        acc[:cost] += (interval_data[:step_rate] || 0) * time_to_charge
        acc[:subsidy] += (interval_data[:step_subsidy] || 0) * time_to_charge
```

@giladshanan what do you think is better? I don't have a strong opinion here